torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -419,172 +450,165 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
+                        global_param_index = self._param_to_index[global_param]
+                        state_dict["state"][global_param_index] = local_state_dict["state"][local_param_index]
+
+        # Sort the parameters in the state
         state_dict["state"] = dict(sorted(state_dict["state"].items()))
         return state_dict
 
     @staticmethod
-    def rank_local_state_dict(rank: int, state_dict: dict) -> dict:
+    def _sync_param_groups(
+        src_param_groups: List[Dict[Any, Any]],
+        dst_param_groups: List[Dict[Any, Any]],
+    ) -> None:
         r"""
-        Returns the local_state_dict for a given rank.
+        Syncs the attributes from the source parameter groups to the
+        destination parameter groups.
+
+        Example attributes include learning rate or scheduler attributes. The
+        two parameter groups should have the same length (i.e. same number of
+        parameter groups).
 
         Arguments:
-            rank (int): rank to get ``local_state_dict`` for
-            state_dict (dict): global ``state_dict``
+            src_param_groups (list[dict]): parameter groups giving the
+                attribute settings to copy.
+            dst_param_groups (list[dict]): parameter groups giving the
+                attribute settings to set.
         """
-        param_groups = state_dict["param_groups"][state_dict["partition"][rank][0] : state_dict["partition"][rank][1]]
-        return {"state": state_dict["state"][rank], "param_groups": param_groups}
+        assert len(src_param_groups) == len(dst_param_groups), \
+            "Mismatch between number of source and destination parameter groups"
+        for src_param_group, dst_param_group in zip(src_param_groups, dst_param_groups):
+            # Sync all attributes except the parameters
+            for attr in filter(lambda x: x != "params", src_param_group.keys()):
+                dst_param_group[attr] = src_param_group[attr]
+
+    def _build_param_buckets(self) -> None:
+        r"""
+        Builds parameter buckets so that for each device that stores this
+        rank's parameters, there is a bucket (represented as a tensor)
+        containing all of the parameters on that device that are assigned to a
+        given rank, if ``parameters_as_bucket_view`` is enabled.
 
-    @staticmethod
-    def _sync_param_groups(source: List[Dict[Any, Any]], destination: List[Dict[Any, Any]]) -> None:
-        r"""Sync learning rate and other optimizer attributes (needed to support schedulers)."""
+        This function is called in the constructor and any time parameter
+        trainability is changed.
 
-        for source_group, destination_group in zip(source, destination):
-            # Sync everything but the parameters
-            for k in filter(lambda x: x != "params", source_group.keys()):
-                destination_group[k] = source_group[k]
+        NOTE: The current implementation assumes that each rank stores all of
```

Comment:
(no need to change this PR, just FYI) if you would like to add a note block, this needs to be `.. note::`. But since this is a private API and the docs won't render, the current one also works. 

https://github.com/pytorch/pytorch/blob/4887c6e401b73c6b024a344076899fd8cdf4a26b/torch/nn/parallel/distributed.py#L216